File: src/generator.cpp
#include "voltorb/generator.hpp"

#include "voltorb/board_type.hpp"

#include <algorithm>
#include <numeric>

namespace voltorb {

BoardGenerator::BoardGenerator(uint64_t seed) : rng_(seed) {
    if (seed == 0) {
        std::random_device rd;
        rng_.seed(rd());
    }
}

void BoardGenerator::seed(uint64_t seed) {
    rng_.seed(seed);
}

bool BoardGenerator::isLegal(const Board& board, const BoardTypeParams& params) const {
    const auto& rowHints = board.rowHints();
    const auto& colHints = board.colHints();

    // Count multipliers in free rows/columns
    int totalFreeMultipliers = 0;

    for (size_t i = 0; i < BOARD_SIZE; i++) {
        for (size_t j = 0; j < BOARD_SIZE; j++) {
            PanelValue v = board.get(static_cast<uint8_t>(i), static_cast<uint8_t>(j));
            if (isMultiplier(v)) {
                if (rowHints[i].voltorbCount == 0 || colHints[j].voltorbCount == 0) {
                    totalFreeMultipliers++;
                }
            }
        }
    }

    if (totalFreeMultipliers >= params.maxTotalFree) {
        return false;
    }

    // Check per-row constraints
    for (size_t i = 0; i < BOARD_SIZE; i++) {
        if (rowHints[i].voltorbCount == 0) {
            int rowMultipliers = 0;
            for (size_t j = 0; j < BOARD_SIZE; j++) {
                if (isMultiplier(board.get(static_cast<uint8_t>(i), static_cast<uint8_t>(j)))) {
                    rowMultipliers++;
                }
            }
            if (rowMultipliers >= params.maxRowFree) {
                return false;
            }
        }
    }

    // Check per-column constraints
    for (size_t j = 0; j < BOARD_SIZE; j++) {
        if (colHints[j].voltorbCount == 0) {
            int colMultipliers = 0;
            for (size_t i = 0; i < BOARD_SIZE; i++) {
                if (isMultiplier(board.get(static_cast<uint8_t>(i), static_cast<uint8_t>(j)))) {
                    colMultipliers++;
                }
            }
            if (colMultipliers >= params.maxRowFree) {
                return false;
            }
        }
    }

    return true;
}

Board BoardGenerator::generate(Level level) {
    std::uniform_int_distribution<int> typeDist(0, NUM_TYPES_PER_LEVEL - 1);
    BoardTypeIndex type = static_cast<BoardTypeIndex>(typeDist(rng_));
    return generate(level, type);
}

Board BoardGenerator::generate(Level level, BoardTypeIndex type) {
    const auto& params = BoardTypeData::params(level, type);

    lastRejectionCount_ = 0;

    while (true) {
        Board board(level);

        // Initialize all panels to 1
        for (size_t i = 0; i < BOARD_SIZE; i++) {
            for (size_t j = 0; j < BOARD_SIZE; j++) {
                board.set(static_cast<uint8_t>(i), static_cast<uint8_t>(j), PanelValue::One);
            }
        }

        // Create shuffled position array
        std::array<size_t, TOTAL_PANELS> positions;
        std::iota(positions.begin(), positions.end(), 0);
        std::shuffle(positions.begin(), positions.end(), rng_);

        size_t idx = 0;

        // Place voltorbs
        for (int n = 0; n < params.n0; n++) {
            Position pos = Position::fromIndex(positions[idx++]);
            board.set(pos, PanelValue::Voltorb);
        }

        // Place 2s
        for (int n = 0; n < params.n2; n++) {
            Position pos = Position::fromIndex(positions[idx++]);
            board.set(pos, PanelValue::Two);
        }

        // Place 3s
        for (int n = 0; n < params.n3; n++) {
            Position pos = Position::fromIndex(positions[idx++]);
            board.set(pos, PanelValue::Three);
        }

        // Calculate hints
        board.recalculateHints();

        // Check legality
        if (isLegal(board, params)) {
            return board;
        }

        lastRejectionCount_++;
    }
}

} // namespace voltorb

File: web/js/generator.js
// Board Generator for Voltorb Flip
// Ported from C++ src/generator.cpp

import {
    BOARD_SIZE,
    TOTAL_PANELS,
    NUM_TYPES_PER_LEVEL,
    PanelValue,
    getParams,
    isMultiplier
} from './boardTypes.js';

import { Board, indexToPos } from './board.js';

// Simple seeded random number generator (Mulberry32)
function mulberry32(seed) {
    return function() {
        let t = seed += 0x6D2B79F5;
        t = Math.imul(t ^ t >>> 15, t | 1);
        t ^= t + Math.imul(t ^ t >>> 7, t | 61);
        return ((t ^ t >>> 14) >>> 0) / 4294967296;
    };
}

export class BoardGenerator {
    constructor(seed = null) {
        this.seed_ = seed || Date.now();
        this.random = mulberry32(this.seed_);
        this.lastRejectionCount_ = 0;
    }

    // Seed the RNG
    seed(seed) {
        this.seed_ = seed;
        this.random = mulberry32(seed);
    }

    // Shuffle array using Fisher-Yates
    shuffle(array) {
        const arr = [...array];
        for (let i = arr.length - 1; i > 0; i--) {
            const j = Math.floor(this.random() * (i + 1));
            [arr[i], arr[j]] = [arr[j], arr[i]];
        }
        return arr;
    }

    // Check if a board configuration is legal
    isLegal(board, params) {
        const rowHints = board.rowHints;
        const colHints = board.colHints;

        // Count multipliers in free rows/columns
        let totalFreeMultipliers = 0;

        for (let i = 0; i < BOARD_SIZE; i++) {
            for (let j = 0; j < BOARD_SIZE; j++) {
                const v = board.get(i, j);
                if (isMultiplier(v)) {
                    if (rowHints[i].voltorbCount === 0 || colHints[j].voltorbCount === 0) {
                        totalFreeMultipliers++;
                    }
                }
            }
        }

        if (totalFreeMultipliers >= params.maxTotalFree) {
            return false;
        }

        // Check per-row constraints
        for (let i = 0; i < BOARD_SIZE; i++) {
            if (rowHints[i].voltorbCount === 0) {
                let rowMultipliers = 0;
                for (let j = 0; j < BOARD_SIZE; j++) {
                    if (isMultiplier(board.get(i, j))) {
                        rowMultipliers++;
                    }
                }
                if (rowMultipliers >= params.maxRowFree) {
                    return false;
                }
            }
        }

        // Check per-column constraints
        for (let j = 0; j < BOARD_SIZE; j++) {
            if (colHints[j].voltorbCount === 0) {
                let colMultipliers = 0;
                for (let i = 0; i < BOARD_SIZE; i++) {
                    if (isMultiplier(board.get(i, j))) {
                        colMultipliers++;
                    }
                }
                if (colMultipliers >= params.maxRowFree) {
                    return false;
                }
            }
        }

        return true;
    }

    // Generate a random board for a given level
    generate(level) {
        const type = Math.floor(this.random() * NUM_TYPES_PER_LEVEL);
        return this.generateWithType(level, type);
    }

    // Generate a random board for a given level and type
    generateWithType(level, type) {
        const params = getParams(level, type);
        this.lastRejectionCount_ = 0;

        while (true) {
            const board = new Board(level);

            // Initialize all panels to 1
            for (let i = 0; i < BOARD_SIZE; i++) {
                for (let j = 0; j < BOARD_SIZE; j++) {
                    board.set(i, j, PanelValue.One);
                }
            }

            // Create shuffled position array
            const positions = [];
            for (let i = 0; i < TOTAL_PANELS; i++) {
                positions.push(i);
            }
            const shuffled = this.shuffle(positions);

            let idx = 0;

            // Place voltorbs
            for (let n = 0; n < params.n0; n++) {
                const pos = indexToPos(shuffled[idx++]);
                board.set(pos.row, pos.col, PanelValue.Voltorb);
            }

            // Place 2s
            for (let n = 0; n < params.n2; n++) {
                const pos = indexToPos(shuffled[idx++]);
                board.set(pos.row, pos.col, PanelValue.Two);
            }

            // Place 3s
            for (let n = 0; n < params.n3; n++) {
                const pos = indexToPos(shuffled[idx++]);
                board.set(pos.row, pos.col, PanelValue.Three);
            }

            // Calculate hints
            board.recalculateHints();

            // Check legality
            if (this.isLegal(board, params)) {
                return board;
            }

            this.lastRejectionCount_++;

            // Safety limit to prevent infinite loop
            if (this.lastRejectionCount_ > 10000) {
                console.warn('Board generation: too many rejections, returning invalid board');
                return board;
            }
        }
    }

    // Get the number of rejections in the last generation
    get lastRejectionCount() {
        return this.lastRejectionCount_;
    }
}

// Create a board from explicit panel values (for testing)
export function createBoardFromValues(level, values, rowHints = null, colHints = null) {
    const board = new Board(level);

    for (let i = 0; i < BOARD_SIZE; i++) {
        for (let j = 0; j < BOARD_SIZE; j++) {
            board.set(i, j, values[i][j]);
        }
    }

    if (rowHints && colHints) {
        for (let i = 0; i < BOARD_SIZE; i++) {
            board.setRowHint(i, rowHints[i]);
            board.setColHint(i, colHints[i]);
        }
    } else {
        board.recalculateHints();
    }

    return board;
}

// Create an empty board with hints (for assistant mode)
export function createEmptyBoardWithHints(level, rowHints, colHints) {
    const board = new Board(level);

    for (let i = 0; i < BOARD_SIZE; i++) {
        board.setRowHint(i, { sum: rowHints[i].sum, voltorbCount: rowHints[i].voltorbCount });
        board.setColHint(i, { sum: colHints[i].sum, voltorbCount: colHints[i].voltorbCount });
    }

    return board;
}
